File: java/src/main/java/com/soi/baekjoon/BOJ_11559.java
package com.soi.baekjoon;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Set;

public class BOJ_11559 {
    private static final int[][] pos = new int[][]{{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    private static final char[][] map = new char[12][6];
    private static Set<int[]> pops = new HashSet<>();
    private static boolean[][] visit = new boolean[12][6];

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        for (int i = 0; i < 12; i++) {
            String line = br.readLine();
            map[i] = line.toCharArray();
        }
        int answer = 0;
        while (true) {
            Set<int[]> pops = puyosToPop();
            if (pops.isEmpty()) break;
            popPuyo();
            organizeMap();
            answer++;
        }
        System.out.println(answer);
    }

    private static Set<int[]> puyosToPop() {
        pops = new HashSet<>();
        visit = new boolean[12][6];
        for (int r = 0; r < 12; r++) {
            for (int c = 0; c < 6; c++) {
                if (map[r][c] == '.' || visit[r][c]) {
                    continue;
                }
                bfs(r, c);
            }
        }
        return pops;
    }

    private static void bfs(int rr, int cc) {
        char color = map[rr][cc];

        Set<int[]> groups = new HashSet<>();
        Queue<int[]> queue = new LinkedList<>();

        groups.add(new int[]{rr, cc});
        queue.offer(new int[]{rr, cc});
        visit[rr][cc] = true;

        while (!queue.isEmpty()) {
            int[] curr = queue.poll();
            for (int[] po : pos) {
                int newRr = curr[0] + po[0];
                int newCc = curr[1] + po[1];
                if (!inRange(newRr, newCc) || visit[newRr][newCc] || map[newRr][newCc] != color) {
                    continue;
                }
                groups.add(new int[]{newRr, newCc});
                queue.offer(new int[]{newRr, newCc});
                visit[newRr][newCc] = true;
            }
        }
        if (groups.size() >= 4) {
            pops.addAll(groups);
        }
    }

    private static void popPuyo() {
        for (int[] puyos : pops) {
            map[puyos[0]][puyos[1]] = '.';
        }
    }

    // 아래로 모든 뿌요들 당기기
    private static void organizeMap() {
        for (int c = 0; c < 6; c++) {
            StringBuilder chars = new StringBuilder();
            for (int r = 0; r < 12; r++) {
                if (map[r][c] == '.') {
                    continue;
                }
                chars.append(map[r][c]);
            }
            chars.insert(0, ".".repeat(12 - chars.length()));
            for (int r = 0; r < 12; r++) {
                map[r][c] = chars.charAt(r);
            }
        }

    }

    private static boolean inRange(int rr, int cc) {
        return rr >= 0 && rr < 12 && cc >= 0 && cc < 6;
    }
}

File: kotlin/src/main/kotlin/com/soi/baekjoon/BOJ_11559.kt
package com.soi.baekjoon

import java.io.BufferedReader
import java.io.InputStreamReader


/**
 * @see <a href= "https://www.acmicpc.net/problem/11559">Puyo Puyo</a>
 */

data class Pos(val r: Int, val c: Int)

private val br = BufferedReader(InputStreamReader(System.`in`))
private val map = Array(12) { CharArray(6) }
private val visit = Array(12) { BooleanArray(6) }

private val dirs = arrayOf(
    intArrayOf(-1, 0),
    intArrayOf(1, 0),
    intArrayOf(0, -1),
    intArrayOf(0, 1)
)

fun main() {
    repeat(12) { r ->
        map[r] = br.readLine().toCharArray()
    }

    var answer = 0

    while (true) {
        val pops = findPops()
        if (pops.isEmpty()) break

        pops.forEach { (r, c) -> map[r][c] = '.' }
        gravity()
        answer++
    }

    println(answer)
}

private fun findPops(): MutableSet<Pos> {
    val pops = mutableSetOf<Pos>()

    for (r in 0 until 12)
        for (c in 0 until 6)
            visit[r][c] = false

    for (r in 0 until 12) {
        for (c in 0 until 6) {
            if (map[r][c] == '.' || visit[r][c]) continue
            bfs(r, c, pops)
        }
    }

    return pops
}

private fun bfs(sr: Int, sc: Int, pops: MutableSet<Pos>) {
    val color = map[sr][sc]
    val group = mutableListOf<Pos>()
    val q = ArrayDeque<Pos>()

    q.add(Pos(sr, sc))
    visit[sr][sc] = true
    group.add(Pos(sr, sc))

    while (q.isNotEmpty()) {
        val (r, c) = q.removeFirst()

        for (d in dirs) {
            val nr = r + d[0]
            val nc = c + d[1]

            if (!inRange(nr, nc) || visit[nr][nc] || map[nr][nc] != color) continue

            visit[nr][nc] = true
            q.add(Pos(nr, nc))
            group.add(Pos(nr, nc))
        }
    }

    if (group.size >= 4) pops.addAll(group)
}

private fun gravity() {
    for (c in 0 until 6) {
        var write = 11

        for (r in 11 downTo 0) {
            if (map[r][c] != '.') {
                map[write--][c] = map[r][c]
            }
        }

        while (write >= 0) {
            map[write--][c] = '.'
        }
    }
}

private fun inRange(r: Int, c: Int) =
    r in 0 until 12 && c in 0 until 6